src/components/Table/utils/RowState.ts

```diff
@@ -10,30 +11,47 @@ import { GridDataRow, KEPT_GROUP, reservedRowKinds, SelectedState } from "src";
 export class RowState {
   /** Our row, only ref observed, so we don't crawl into GraphQL fragments. */
   row: GridDataRow<any>;
-  /** Our parent RowState, or the `header` RowState if we're a top-level row. */
-  parent: RowState | undefined;
   /** Our children row states, as of the latest `props.rows`, without any filtering applied. */
   children: RowState[] | undefined = undefined;
   /** Whether we match a client-side filter; true if no filter is in place. */
   isMatched = true;
   /** Whether we are *directly* selected. */
   selected = false;
-  /** Whether our `row` had been in `props.rows`, but was removed, i.e. probably by server-side filters. */
-  wasRemoved = false;
+  /** Whether we are collapsed. */
+  collapsed = false;
+  /**
+   * Whether our `row` had been in `props.rows`, but then removed _while being
+   * selected_, i.e. potentially by server-side filters.
+   *
+   * We have had a large foot-gun where users "select a row", change the filters,
+   * the row disappears (filtered out), and the user clicks "Go!", but the table
+   * thinks their previously-selected row is gone (b/c it's not in view), and
+   * then the row is inappropriately deleted/unassociated/etc. (b/c in the user's
+   * head, it is "still selected").
+   *
+   * To avoid this, we by default keep selected rows, as "kept rows", to make
+   * extra sure the user wants them to go away.
+   *
+   * Soft-deleted rows are rows that were removed from `props.rows` (i.e. we
+   * suspect are just hidden by a changed server-side-filter), and hard-deleted
+   * rows are rows the page called `api.deleteRow` and confirmed it should be
+   * actively removed.
+   */
+  removed: false | "soft" | "hard" = false;
```

Comment:
I should have kept this as a separate PR, but I changed `wasRemoved` to a tri-state, so that `api.deleteRows` can mark the hard-deletion + allowed removing the `RowState.parent` field.